### ISSUE TITLE
Using an `\u000d` escape in a comment "scrambles" the output

### ISSUE THREAD

**Issue by mihnita** [google/error-prone#5031]
When there is a `\u000d` in a block comment `UnicodeEscape` reports the problem twice, in both cases incorrectly.

**Comment by cushon** [comment]
Thanks for the report, could you share an example?

**Comment by mihnita** [comment]
Details, how to reproduce.

---

Code:

```java
public class App {
    /* Line 1\u000dSimple escape in comment: \u0041. Backslash-u-000d. Reported incorrectly. */
    public static void main(String[] args) {
        System.out.println("Hello world!");
    }
}
```

Running errorprone (inspired by the installation page):

```sh
EP_VERSION=2.38.0
DATAFLOW_VERSION=3.49.1-eisop1

if [ ! -f error_prone_core-${EP_VERSION?}-with-dependencies.jar ]; then
  echo Downloading error_prone_core
  wget -q https://repo1.maven.org/maven2/com/google/errorprone/error_prone_core/${EP_VERSION?}/error_prone_core-${EP_VERSION?}-with-dependencies.jar
fi

if [ ! -f dataflow-errorprone-${DATAFLOW_VERSION?}.jar ]; then
  echo Downloading dataflow-errorprone
  wget -q https://repo1.maven.org/maven2/io/github/eisop/dataflow-errorprone/${DATAFLOW_VERSION?}/dataflow-errorprone-${DATAFLOW_VERSION?}.jar
fi

javac \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED \
  -J--add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED \
  -J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED \
  -XDcompilePolicy=simple \
  --should-stop=ifError=FLOW \
  -processorpath error_prone_core-${EP_VERSION?}-with-dependencies.jar:dataflow-errorprone-${DATAFLOW_VERSION?}.jar \
  '-Xplugin:ErrorProne' \
  src/main/java/com/example/App.java
```

The Unicode escape in line 4 is reported twice, both reports somewhat incorrect:

```
src/main/java/com/example/App.java:4: warning: [UnicodeEscape] Using unicode escape sequences for printable ASCII characters is obfuscated, and potentially dangerous.
    /* Line 1\u000dSimple escape in comment: \u0041. Backslash-u-000d. Reported incorrectly. */
                  ^
    (see https://errorprone.info/bugpattern/UnicodeEscape)
Simple escape in comment: \u0041. Backslash-u-000d. Reported incorrectly. */'?
src/main/java/com/example/App.java:4: warning: [UnicodeEscape] Using unicode escape sequences for printable ASCII characters is obfuscated, and potentially dangerous.
    /* Line 1\u000dSimple escape in comment: \u0041. Backslash-u-000d. Reported incorrectly. */
                                                  ^
    (see https://errorprone.info/bugpattern/UnicodeEscape)
  Did you mean '/* Line 1\u000dSimple escape in comment: A. Backslash-u-000d. Reported incorrectly. */'?
2 warnings
```

The first report shows the first line with the file path + location + errorprone code + description.  
Correct and consistent with how other problems are reported.

But the "Did you mean" part drops part of the comment (the one after CR).

---

The second report shows the comment itself as the first line.  
Incorrect.

Then the path + location + errorprone report.

And finally the "Did you mean" part looks correct.

---

I tried using other characters instead of `\u000d`: `\u000a`, `\r`, `\n`, `\r\n`, real newline.  
All of them are reported correctly (only once, with file path + location in the WARNING line).  
But sorry, I did not pay attention to the "Did you mean" line for all of these.

---

Context:

Java version: 21.0.3, vendor: Eclipse Adoptium, runtime: /Users/test/apps/jdk-21.0.3+9/Contents/Home  
Default locale: en_US, platform encoding: UTF-8  
OS name: "mac os x", version: "15.4.1", arch: "aarch64", family: "mac"

I suspect this behaves the same on Linux and maybe on Windows.  
But I can test it if you want me to.

**Comment by mihnita** [comment]
Here is a small maven project (but also includes a `cli.sh` doing the download and the direct javac run)

https://mihai-nita.net/tmp/errprone_bug.tgz

The maven output makes it a bit more clear what's going on:
```
[WARNING] /Users/test/errprone/src/main/java/com/example/App.java:[4,18] [UnicodeEscape] Using unicode escape sequences for printable ASCII characters is obfuscated, and potentially dangerous.
    (see https://errorprone.info/bugpattern/UnicodeEscape)
  Did you mean '/* Line 1
[WARNING] Simple escape in comment: \u0041. Backslash-u-000d. Reported incorrectly. */'?
/Users/test/errprone/src/main/java/com/example/App.java:[4,50] [UnicodeEscape] Using unicode escape sequences for printable ASCII characters is obfuscated, and potentially dangerous.
    (see https://errorprone.info/bugpattern/UnicodeEscape)
  Did you mean '/* Line 1\u000dSimple escape in comment: A. Backslash-u-000d. Reported incorrectly. */'?
```

**Comment by mihnita** [comment]
Same problem on Linux and Windows.

**Comment by graememorgan** [comment]
There are two findings on your example with `Line 1\u000dSimple escape in comment: \u0041. Backslash-u-000d. Reported incorrectly.` because there are two unicode escapes. Fixes for each of those are issued independently.

> The second report shows the comment itself as the first line.
> Incorrect.

I'm not following what's incorrect about that.

> But the "Did you mean" part drops part of the comment (the one after CR).

I think this is expected given the way "Did you mean" is formatted. `0D` is a carriage return, so the fix involves creating a new line. The comment is now two lines, and `AppliedFix` has some logic to try to pick an interesting line to report with "Did you mean ...?". That's obviously not great when the finding consists of adding a carriage return, but if you apply the fixes all should be fine.

**Comment by mihnita** [comment]
OK, I get the two escapes - two messages.

But they "overlap", making the result un-parseable (ok, relatively).

Look at the maven output.

Or, better, don't look at it, grep it:
```
mvn clean compile -P errorprone | grep WARNING
```

The result is:
```
[WARNING] /Users/test/errprone/src/main/java/com/example/App.java:[4,18] [UnicodeEscape] Using unicode escape sequences for printable ASCII characters is obfuscated, and potentially dangerous.
[WARNING] Simple escape in comment: \u0041. Backslash-u-000d. Reported incorrectly. */'?
```

See the second WARNING line.

The expectation is the the `[WARNING]` line is the start of a new reported issue.  
And after the `[WARNING]` there should be the file path, location, error type, description.  
But instead we get the second line from the first message.

And the fact that this also happens on grep seems to indicate that this is not only a visual artifact of how the terminal works, some kind of carriage-return making the WARNING override the previous message without a newline.

---

In fact I found this error by writing a program to scan a maven build result in order to generate something better.
Because errorprone has no option to generate a more structured report we are forced to parse standard outputs.

---

Thank you,
M.

**Comment by graememorgan** [comment]
> The expectation is the the [WARNING] line is the start of a new reported issue.

Is that a documented expectation by Maven?

I poked around and found that you're [parsing Maven outputs into reports](https://github.com/unicode-org/icu/commit/4b444321f255076d6576e6d435ffe2589bde6f18). I can understand why this formatting would be annoying, but I'm not exactly sure it's wrong. I'm not sure how we'd sensibly render a fix that consists of introducing a newline.

There's some docs at https://errorprone.info/docs/patching about using ErrorProne to generate _patches_. I wonder if it would be productive to use that kind of approach to emit genuinely structured finding data? (I don't think we have anything existing in the OSS version, but cushon@ may have ideas.)

**Comment by mihnita** [comment]
> Is that a documented expectation by Maven?

No, but that is how all the other issues reported by errorprone work.

> I'm not sure how we'd sensibly render a fix that consists of introducing a newline
Not mixing it with the WARNING, which marks the beginning of the next issue.

The main problem is that errorprone has no way to produce a report in a structured format.
So we have to scan the standard output from various tools.
If there was a way to generate an xml / json / csv / whatever, we wouldn't need to do this.

There are several feature request issues about that.

It might be that maven itself is confused by the errorprone standard output.

**Comment by mihnita** [comment]
I tried pathing, it does not help.
The patch only contains the difference.
And I need to explicitly list all issues to fix, it does not seem to be not documented way to say "ALL" (I tried `-XepPatchChecks:ALL` and `-XepPatchChecks:*`, I can probably dig in the code).

But I need the issue type, severity level, a description of the issue, and a link to the explanation.
What is currently available in `com.google.errorprone.BugCheckerInfo`

---

Here is some context on what I'm trying to do, to avoid the XY-problem.

We want to generate an errorprone report with all the issues detected in ICU.
Then we want to analyze it, decide what to fix and what not, and priorities.
The (over time) reduce the number of issues (to zero if possible).
And finally to generate such a report for each PR and reject PRs that introduce new issues.
(ideally on the modified lines only).

So it is not so much a "fix it" effort (although some of that will happen) as it is a "make sure we don't introduce problems" in the long run.

Problems I've encountered trying to do that:
- Needed to parse the Maven standard output. Other people are forced to parse standard output from gradle, ant, cli (duplicated work)
- The output format is not documented, and it goes to standard output.
So is mixed with all kind of other issues from the compiler, other tools, possibly print statements in tests, and log entries from test code or from Maven.
- An error means that the compilation stops. So there is no way to generate a full report with all issues. I was forced to convert all errors to warnings. But then I lost the distinction between really important issues (errors) vs the not so important ones (warnings). So needed to add a dependency on errorprone, access the list of all know BugCheckerInfo, and get the default severity). Which also means it is harder to write such a report generating tool in something other than Java.

See here for the two kind of reports I generate:
- https://mihai-nita.net/tmp/ep/errorprone1.html
- https://mihai-nita.net/tmp/ep/errorprone2.html

---

A report in a structured file format, as requested in several open issues, would really help.
There would be no need for all of this "gymnastics".

I offer to implement it, following whatever guidelines you set, whatever format you prefer.

**Comment by cushon** [comment]
There's a previous issue about structured diagnostic output in #3766

**Comment by mihnita** [comment]
Yes. And here: #444 (where I also offered to contribute that functionality :-)

**Comment by cushon** [comment]
I think the diagnostic output here is basically WAI for the reasons in https://github.com/google/error-prone/issues/5031#issuecomment-2854115575

For the discussion of structured diagnostics, let's continue in #444 / #3766
